javaslang/src-gen/test/java/javaslang/CheckedFunction2Test.java

```diff
@@ -165,15 +165,6 @@ public void shouldLiftTryPartialFunction() {
         assertThat(unknown.getCause().getMessage()).isEqualToIgnoringCase("Unknown MessageDigest not available");
     }
 
-    private static final CheckedFunction2<Integer, Integer, Integer> recurrent1 = (i1, i2) -> i1 <= 0 ? i1 : CheckedFunction2Test.recurrent2.apply(i1 - 1, i2) + 1;
```

Comment:
I'm concerned around
```
if (cache.containsKey(key)) {
    return cache.get(key);
} else {
    synchronized (cache) {
        ...
    }
}
```

It seems an awful lot to me like previous discussion on `Lazy`. The key idea: the _then_ branch above doesn't use any volatile variable, neither does it pass through a `synchronized` block, nor does `cache` involve any volatile reads inside as it's a simple `HashMap`. Because no memory barrier is passed, it is unsafe. See [Pitfall: Semi-Synchronized Is Fine](https://shipilev.net/blog/2016/close-encounters-of-jmm-kind/#pitfall-semi-sync).